scipy/signal/_peak_finding.py

```diff
@@ -233,6 +241,578 @@ def argrelextrema(data, comparator, axis=0, order=1, mode='clip'):
+    >>> peaks
+    array([ 57, 306, 556, 806])
+
+    Plot results
+
+    >>> import matplotlib.pyplot as plt
+    >>> countour_height = vector[peaks] - info['prominences']
+    >>> plt.plot(time, vector)
+    >>> plt.plot(time[peaks], vector[peaks], "x")
+    >>> plt.vlines(time[peaks], ymin=countour_height, ymax=vector[peaks])
+    """
+    vector = np.asarray(vector)
+    peaks = argrelmax(vector)[0]
+    infodict = {}
+
+    # Filter by height
+    if height is not None:
+        hmin, hmax = _parse_filter_args(height, vector, peaks)
+        heights = vector[peaks]
+        keep = np.ones(peaks.size, dtype=bool)
+        if hmin is not None:
+            keep &= (vector[peaks] >= hmin)
+        if hmax is not None:
+            keep &= (vector[peaks] <= hmax)
+        peaks = peaks[keep]
+        if full_output:
+            infodict["heights"] = heights[keep]
+
+    # Filter by threshold
+    if threshold is not None:
+        tmin, tmax = _parse_filter_args(threshold, vector, peaks)
+        keep, stacked_thresholds = _filter_by_peak_threshold(
+            vector, peaks, tmin, tmax)
+        peaks = peaks[keep]
+        if full_output:
+            infodict["stacked_thresholds"] = stacked_thresholds
+            infodict = {key: array[keep] for key, array in infodict.items()}
+
+    # Filter by distance
+    if distance is not None:
+        keep = _filter_by_peak_distance(vector, peaks, distance)
+        peaks = peaks[keep]
+        if full_output:
+            infodict = {key: array[keep] for key, array in infodict.items()}
+
+    if prominence is not None or width is not None:
+        # Prominence is required for both filter options
+        prominences = peak_prominences(vector, peaks, wlen=wlen)
+
+        # Filter by prominence
+        if prominence is not None:
+            pmin, pmax = _parse_filter_args(prominence, vector, peaks)
+            keep = np.ones(peaks.size, dtype=bool)
+            if pmin is not None:
+                keep &= (pmin <= prominences)
+            if pmax is not None:
+                keep &= (prominences <= pmax)
+            peaks = peaks[keep]
+            prominences = prominences[keep]
+            if full_output:
+                infodict = {key: array[keep] for key, array in infodict.items()}
+                # Don't remove peaks again
+                infodict["prominences"] = prominences
+
+        # Filter by width
+        if width is not None:
+            wheight = width[0]
+            wmin, wmax = _parse_filter_args(width[1:], vector, peaks)
+            widths, wheights, lpos, rpos = peak_widths(
+                vector, peaks, wheight, prominences, ret_pos=True)
+            keep = np.ones(peaks.size, dtype=bool)
+            if wmin is not None:
+                keep &= (wmin <= widths)
+            if wmax is not None:
+                keep &= (widths <= wmax)
+            peaks = peaks[keep]
+            if full_output:
```

Comment:
I almost wonder if the dictionary data structure should just be used directly for storing values, if the extra return data remains part of the API.

Reply:
Do you suggest using, for example, `infodict['peaks']` instead of `peaks`? I think that would be possible. In that case infodict would always be used internally but only returned if `full_output is True`. However I think the trade off would be a slight decrease in readability because the line length would grow significantly...

Note: I have copied this style from [scipy.integrate.odeint](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.odeint.html).

Reply:
Yeah, something like that. I just wanted to make sure the point was brought up & trade-offs at least briefly considered. It looks like there was already quite some discussion of API matters in the original issue you linked.